### PR TITLE
Upstream/merge v3.1.0

### LINKED ISSUE #17: Upgrade IBC-Go to v0.3.1
## Summary
Make the repo in sync with the original lib, Upgrade and merge the code.

## Problem Definition
The new version of upstream [lib](https://github.com/cosmos/ibc-go/releases/tag/v3.1.0) is released.

## Proposal
This repo should be upgraded to the latest version code-base.

### PR DESCRIPTION
**Pull request by TxCorpi0x**
## Description
Upgrade code according the upstream cosmos [IBC-Go](https://github.com/cosmos/ibc-go/tree/v3.1.0).

closes: #17 

---

- [x] Import latest changes from upstream repo
- [x] Check trie new version and `VerifyProof` method replacement
- [ ] Implement `CheckHeaderAndUpdateState` Method of the client state for beefy

### CODE REVIEW

File: modules/apps/27-interchain-accounts/host/types/keys.go [hunk 1]
```diff
@@ -14,6 +14,11 @@ const (
 
 // ContainsMsgType returns true if the sdk.Msg TypeURL is present in allowMsgs, otherwise false
 func ContainsMsgType(allowMsgs []string, msg sdk.Msg) bool {
+	// check that wildcard * option for allowing all message types is the only string in the array, if so, return true
+	if len(allowMsgs) == 1 && allowMsgs[0] == "*" {
```
Comment (blasrodri) [inline modules/apps/27-interchain-accounts/host/types/keys.go:18]: alternatively use https://pkg.go.dev/strings#EqualFold

File: modules/apps/transfer/types/trace.go [hunk 2]
```diff
@@ -70,6 +74,24 @@ func (dt DenomTrace) GetFullDenomPath() string {
...
 }
 
+// extractPathAndBaseFromFullDenom returns the trace path and the base denom from
+// the elements that constitute the complete denom.
+func extractPathAndBaseFromFullDenom(fullDenomItems []string) (string, string) {
+	var path []string
+	var baseDenom []string
+	length := len(fullDenomItems)
+	for i := 0; i < length; i = i + 2 {
+		if i < length-1 && length > 2 && fullDenomItems[i] == PortID {
```
Comment (blasrodri) [inline modules/apps/transfer/types/trace.go:84]: it would be nice to get some comments here on what these conditions imply

**Review by seunlanlege — CHANGES_REQUESTED** [review]
Naahhh we can't merge this.

**Comment by seunlanlege** [comment]
Our work should be based on this branch: https://github.com/cosmos/ibc-go/tree/02-client-refactor

**Pull request closed without merging**